### PR TITLE
[1.1] Fix set nofile rlimit error

### PR DESCRIPTION
**Pull request by lifubang**
This is backport of #4268, #4265.
------
Fix: #4195 
Close: #4237 

## 1. Fix a get/set race between `runc exec` and `syscall.rlimit.init()`
As @ls-ggg has given the detailed steps to reproduce the issue #4195 , and has given the core reason is that the go runtime will cache the value of rlimit nofile. [1] 
When we are running `runc exec` with nofile limit, runc set it in the parent process, it may cause a race with go runtime init. The race condition is in the time when runc parent process set the container child process's nofile limit just after go runtime init has fetched the nofile limit. [2]
So, we should set nofile limit after `syscall.rlimit.init()` completed.

## 2. Fix an edge case caused by nofile rlimit cache in go stdlib
As @kolyshkin have found an edge case for set nofile rlimit by runc, which is also caused by nofile rlimit cache in go stdlib. Although we have a way to resolve the above get/set race, but if the hard value of nofile rlimit configured in `config.json` is bigger than this value of `runc create/run/exec`, it will also be incorrect restored by `syscall.Exec` in `runc init`. So we need to clear the nofile rlimit cache before we start the container initial process if we need to set nofile rlimit for the container.

[1] https://github.com/golang/go/commit/f5eef58e4381259cbd84b3f2074c79607fb5c821#diff-ec665e9789f8cf5cd1828ad7fa9f0ff4ebc1f5b5dd0fc82a296da5c07da7ece6
[2] https://github.com/golang/go/blob/f5eef58e4381259cbd84b3f2074c79607fb5c821/src/syscall/rlimit.go#L34-L35

----
CHANGELOG:
deprecated: libct.system.Execv is not used in runc anymore, it will be removed in v1.2.0.

### CODE REVIEW

**Comment by lifubang** [comment]
Note that we don’t support go version less than 1.19 anymore once we merge this PR.

**Comment by kolyshkin** [comment]
> Note that we don’t support go version less than 1.19 anymore once we merge this PR.

This is a questionable decision. I think what we should do instead in 1.1 is add two versions of `func ClearRlimitNofileCache` -- one for Go 1.19+ which does what it does now, and the other is for !go1.19, which does nothing.

**Comment by lifubang** [comment]
> I think what we should do instead in 1.1 is add two versions of `func ClearRlimitNofileCache` -- one for Go 1.19+ which does what it does now, and the other is for `!go1.19`, which does nothing.

Thanks your suggestion.

File: libcontainer/system/rlimit_stub.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+package system
+
+// As reported in issue #4195, the new version of go runtime(since 1.19)
+// will cache rlimit-nofile. Before executing execve, the rlimit-nofile
+// of the process will be restored with the cache. In runc, this will
+// cause the rlimit-nofile setting by the parent process for the container
+// to become invalid. It can be solved by clearing this cache. But
+// unfortunately, go stdlib doesn't provide such function, so we need to
+// link to the private var `origRlimitNofile` in package syscall to hack.
```
Comment (kolyshkin) [inline libcontainer/system/rlimit_stub.go:11]: No need to repeat the comment I guess.

**Comment by lifubang** [comment]
> The complete generic features are supported since go 1.18

The complete generic features are supported since `go 1.18`, because go 1.19 and above need this change.

**Comment by kolyshkin** [comment]
> > The complete generic features are supported since go 1.18
> 
> The complete generic features are supported since `go 1.18`, because go 1.19 and above need this change.

I'm sorry, I still don't understand what do you mean :(

This code works with Go 1.17 AFAICS. Can you explain why you changed go version to 1.18 in go.mod in the last commit?

**Comment by lifubang** [comment]
> This code works with Go 1.17 AFAICS.

Yes, this works with go 1.17, but if we use go 1.19, because of the ‘go 1.17’ defined in ‘go.mod’, it can’t work.

**Comment by kolyshkin** [comment]
Ah, I see! This is what happens without the commit bumping the go version in go.mod:

```console
[kir@kir-rhat runc]$ make all GO=go1.20
go1.20 build -trimpath "-buildmode=pie"  -tags "seccomp" -ldflags "-X main.gitCommit=v1.1.12-21-ge36b3ae7 -X main.version=1.1.12+dev " -o runc .
# github.com/opencontainers/runc/libcontainer/system
libcontainer/system/rlimit_go119.go:13:43: type instantiation requires go1.18 or later (-lang was set to go1.17; check go.mod)
make: *** [Makefile:61: runc] Error 1
```

So, despite `libcontainer/system/rlimit_go119.go` being guarded with `//go:build go1.19`, go build still complains. To me this looks like a bug, and there's an issue opened about this (https://github.com/golang/go/issues/52880), but it looks like it won't be fixed.

This is why the go version bump is needed.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
> So, despite libcontainer/system/rlimit_go119.go being guarded with //go:build go1.19, go build still complains. To me this looks like a bug, and there's an issue opened about this (golang/go#52880), but it looks like it won't be fixed.
> 
> This is why the go version bump is needed.


Wait, so we're using go1.20, but the go version in go.mod causes it to downgrade to _language version 1.17_, but ... then it does _not_ use the < go1.19 fallback (through `//go:build` constraints) implementation? So what's the value of `go` build-constraints now if the only version used is the version specified in `go.mod`, and consumers of the module MUST be using that version (or above)? Does that mean it's no longer possible to provide fallback implementations?

This stuff is getting more and more confusing.

File: libcontainer/system/rlimit_go119.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	"sync/atomic"
+	"syscall"
+
+	_ "unsafe" // for go:linkname
+)
+
+//go:linkname syscallOrigRlimitNofile syscall.origRlimitNofile
+var syscallOrigRlimitNofile atomic.Pointer[syscall.Rlimit]
+
+// As reported in issue #4195, the new version of go runtime(since 1.19)
```
Comment (thaJeztah) [inline libcontainer/system/rlimit_go119.go:15]: Looks like this was also in the cherry-picked commit, but we should fix this comment to be in correct GoDoc format (on main at least)

**Review by thaJeztah — COMMENTED** [review]
Curious why we're back porting this one;

- https://github.com/opencontainers/runc/pull/4268

Technically that's a breaking change so not something were could include in a patch release. Perhaps we're not using it ourselves (so it would be "dead code" for us), but someone using the module _could_ be using it.

I couldn't find external consumers, but if we have no strong reason, perhaps we should skip it to reduce the diff for path releases?

**Comment by lifubang** [comment]
> Curious why we're back porting this one;
> 
> * [Remove unused system.Execv #4268](https://github.com/opencontainers/runc/pull/4268)
> 
> Technically that's a breaking change so not something were could include in a patch release. Perhaps we're not using it ourselves (so it would be "dead code" for us), but someone using the module _could_ be using it.
> 
> I couldn't find external consumers, but if we have no strong reason, perhaps we should skip it to reduce the diff for path releases?

I agree. In fact we have no strong reason to back port this commit, we just only need a comment change from this commit.
I'll remove this backport later.
Thanks.

**Comment by thaJeztah** [comment]
^ I'd be fine adding a `Deprecated: ` comment on it (for the 1.1 branch) if we want to raise visibility for it going away. I think there's some other exported functions that we removed in 1.2 as well, and I guess every bit could help to raise awareness.

**Comment by lifubang** [comment]
> Wait, so we're using go1.20, but the go version in go.mod causes it to downgrade to _language version 1.17_, but ... then it does _not_ use the < go1.19 fallback (through `//go:build` constraints) implementation?

No, because go version still be go1.20.

There are two explanations related to go directive in go.mod:
1. The go directive affects use of new language features
This means that if we using go 1.19 to compile runc, we can’t use any new language features provided after 1.17, for example generic type;

2. Before Go 1.21, the directive was advisory only
This means that we can support go 1.17 like before.

Refer:
https://go.dev/doc/modules/gomod-ref#go

**Comment by lifubang** [comment]
> ^ I'd be fine adding a `Deprecated: ` comment on it (for the 1.1 branch) if we want to raise visibility for it going away. I think there's some other exported functions that we removed in 1.2 as well, and I guess every bit could help to raise awareness.

Switched to this suggection, make it deprecated in 1.1 .

**Comment by kolyshkin** [comment]
A single nit: I think `libct: clean cached rlimit nofile in go runtime` should come AFTER `use go 1.18 in go.mod`, as otherwise we're breaking git-bisect.

**Comment by lifubang** [comment]
> A single nit: I think `libct: clean cached rlimit nofile in go runtime` should come AFTER `use go 1.18 in go.mod`, as otherwise we're breaking git-bisect.

done

File: go.mod [hunk 3]
```diff
@@ -1,6 +1,6 @@
 module github.com/opencontainers/runc
 
-go 1.17
+go 1.18
```
Comment (AkihiroSuda) [inline go.mod:3]: (Ideally we should not bump up Go version in backport releases, but it's not a huge deal, as Go 1.17 has already reached EOL)

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**